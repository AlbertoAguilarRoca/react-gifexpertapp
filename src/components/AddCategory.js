import React, { useState } from 'react';
import PropTypes from 'prop-types';

//recibe por parametro la funcion del hook useState
export const AddCategory = ( {setCategories} ) => {
    //hook useState con la informacion del input 
    const [inputValue, setinputValue] = useState('');

    const handleInputValue = (e) => {
        setinputValue(e.target.value);
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        
        //metemos dentro del useState la nueva categoria
        if(inputValue.trim().length > 2) {
            //Metemos el nuevo valor junto con las categorias
            //que ya habia antes
            setCategories(category => [inputValue, ...category ]);
            setinputValue('');
        }
    }

    return (
        <form onSubmit={handleSubmit}>
            <div className='buscador'>
                <i className='bi bi-plus-circle'></i>
                <input 
                    type='text' 
                    value={ inputValue }
                    onChange = { handleInputValue }
                />
            </div>
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
