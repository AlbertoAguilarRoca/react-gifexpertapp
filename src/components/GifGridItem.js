import React from 'react'
import PropTypes from 'prop-types';

export const GifGridItem = ({title, url}) => {

  return (
    <div className='card animate__animated animate__fadeInDown'>
      <div className='card__content'>
        <img src={url} alt={title} />
        <h3>{title}</h3>
      </div>
    </div>
  )
}

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired
}
