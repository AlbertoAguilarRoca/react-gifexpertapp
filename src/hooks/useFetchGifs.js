import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs";



export const useFetchGifs = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    //nuevo hook para ejecutar solo una vez la funcion getGifs()
    useEffect(() => {
        getGifs(category)
            .then(imgs => {
                setState({
                    data: imgs,
                    loading: false
                })
            })

    }, [category]);

    return state;
}