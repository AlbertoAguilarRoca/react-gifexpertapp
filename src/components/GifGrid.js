import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem'

//se recibe la categoria que se va a buscar
export const GifGrid = ({category}) => {

    const {data:images, loading} = useFetchGifs(category);

    //imprimira las imagenes a traves de un nuevo componente
    //en vez de enviar un parametro, se envia ...img para poder
    //desestructurar dicho componente
    //Cuando el componente se ejecuta varias veces, hay que poner una key unica
    return (
        <>
            <h2 className='animate__animated animate__fadeInDown'>{category}</h2>

            {loading && <p className='animate__animated animate__fadeInDown'>
                Cargando...</p>}

            <div className='card__row'>
                {
                    images.map( img => (
                        <GifGridItem 
                            key = {img.id}
                            {...img} 
                        />
                    ))
                }
            </div>
        </>
    )
}
