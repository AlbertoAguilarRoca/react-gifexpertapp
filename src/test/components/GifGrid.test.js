import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Test de <GifGrid />', () => {
    const category = 'One Punch';

    test('debe de mostrar correctamente <GifGrid />', () => {  
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrid category={category} />);
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de mostrar items cuando se cargan imagenes useFetchGifs', () => {
        const gif = [{
            id: 'abc',
            url: 'https://localhost/cosa.jpg',
            title: 'Prueba'
        }];

        useFetchGifs.mockReturnValue({
            data: gif,
            loading: false
        });
        const wrapper = shallow(<GifGrid category={category} />);

        expect(wrapper.find('p').exists()).toBe(false);
        /*Evaluando si existe un componente*/
        expect(wrapper.find('GifGridItem').length).toBe(gif.length);
    });
});