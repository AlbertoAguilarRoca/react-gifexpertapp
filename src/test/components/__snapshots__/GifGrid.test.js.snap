// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test de <GifGrid /> debe de mostrar correctamente <GifGrid /> 1`] = `
<Fragment>
  <h2
    className="animate__animated animate__fadeInDown"
  >
    One Punch
  </h2>
  <p
    className="animate__animated animate__fadeInDown"
  >
    Cargando...
  </p>
  <div
    className="card__row"
  />
</Fragment>
`;
